Add a new chapter to define concepts related to pattern-matching. (#757)
dotnet/csharpstandard: standard/patterns.md

> I would suggest placing this new chapter between the existing 12 (Statements) and 13 (namespaces)

Reviewer: Another placement would be before the existing 11 (Expressions). We tend to introduce building blocks before we use them, and patterns are used by expressions and statements. 

Another option is to not make them a top level chapter but add them to the set of sections in the beginning of Expressions providing building blocks for the subsequent syntax-based sections. This is what I have proposed to do with "Deconstruction" in #644, and pattern matching in a later version of C# will make use of deconstruction, so having patterns follow deconstruction would fit with that.

Author: Before expressions makes sense to me.  I suggested after only because expressions seem like a more fundamental and important concept, and each uses the other.

The patterns "chapter" will get more complicated when we add more pattern concepts in the next couple of versions (and in the future if discriminated unions are added), so I wanted to break them out so as not to interrupt the flow of the expressions description.

# §patterns-new-clause Patterns and pattern matching

## §patterns-new-clause-general General

> A ***pattern*** is a syntactic form that can be used with the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)) and in a *switch_statement* ([§12.8.3](statements.md#1283-the-switch-statement)) to express the shape of data against which incoming data is to be compared. A pattern is tested in the context of a switch expression or a *relational_expression* that is on the left-hand side of an `is` operator. We call this a ***pattern input value***.

Reviewer: ```suggestion
A ***pattern*** is a syntactic form that can be used with the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)) and in a *switch_statement* ([§12.8.3](statements.md#1283-the-switch-statement)) to express the shape of data against which incoming data is to be compared. A pattern is tested against the *expression* of a switch statement, or against a *relational_expression* that is on the left-hand side of an `is` operator. We call this a ***pattern input value***.
```

## §patterns-new-clause-forms Pattern Forms

A pattern may have one of the following forms:

```ANTLR
pattern
    : declaration_pattern
    | constant_pattern
    | var_pattern
    ;
```

A *declaration_pattern* and a *var_pattern* can result in the declaration of a local variable.

> Each pattern form defines the set of types for input values that that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.

Reviewer: ```suggestion
Each pattern form defines the set of types for input values that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.
```

Each pattern form defines the set of values for which the pattern *matches* the value.

### §declaration-pattern-new-clause Declaration pattern

> A *declaration_pattern* is used to test that a value has a given type and if the test succeeds, to cast that value to that type.

Reviewer: The term "cast" seems potentially suggestive of different behavior.

```suggestion
A *declaration_pattern* is used to test that a value has a given type and, if the test succeeds, provide the value in a variable of that type.
```

```ANTLR
declaration_pattern
    : type simple_designation
    ;

> simple_designation

Reviewer: Remind me why we have this chain of trivial productions - is it in anticipation of future additions?

Author: Yes, we have tuple-like designations in the future, and it becomes optional.

    : single_variable_designation
    ;
single_variable_designation
    : identifier
    ;
```

> The runtime type of the value is tested against the *type* in the pattern. If it is of that runtime type (or some subtype), the pattern *matches* that value. This pattern form never matches a `null` value.

Reviewer: > If it is of that runtime type (or some subtype), the pattern *matches* that value.

This seems underspecified. Compare to the detail in the specification of the existing type-based `is` operator. In fact I'd suggest moving all that detail here and respecifying the `E is T` operator as simply a shorthand for `E is T _`.

Author: The is-type operator permits a nullable type on the right-hand-side.  The is-pattern operator does not.  But we can still unify most of it.

Reviewer 2: Also should "subtype" be "pattern-compatilble type"?

Reviewer 3: Addressed in #934 

reworded to use the rules in `is type`, with the added rule that the type can't be a nullable value type.

> The runtime type of the value is tested against the *type* in the pattern. If it is of that runtime type (or some subtype), the pattern *matches* that value. This pattern form never matches a `null` value.

Reviewer 3: This should include boxing and unboxing conversions, right?

Author: Yes.  We should unify with the description of the is-type operator.

Author: This section should refer to the specification of the is-type operator for the meaning of when it is true or false.  This is probably better done sooner rather than later.

Reviewer 3: Done in #934

> Given a *pattern input value* *e*, if the *simple_designation* is the *identifier* `_`, it denotes a discard (§discards-new-clause) the value of *e* is not bound to anything. (Although a declared variable with the name `_` may be in scope at that point, that named variable is not seen in this context.) If *simple_designation* is any other identifier, a local variable ([§9.2.8](variables.md#928-local-variables)) of the given type named by the given identifier is introduced. That local variable is assigned the value of the *pattern input value* when the pattern *matches* the value.

Reviewer: This may need a bit of tinkering once this PR meets up with the discards coming in from #664.

Author: Maybe, maybe not.  I suggest creating an issue to examine this in detail at some later time.  I suspect no action is needed.

Reviewer 3: Logged in #936

Certain combinations of static type of the pattern input value and the given type are considered incompatible and result in a compile-time error. A value of static type `E` is said to be ***pattern compatible*** with the type `T` if there exists an identity conversion, an implicit reference conversion, a boxing conversion, an explicit reference conversion, or an unboxing conversion from `E` to `T`, or if either `E` or `T` is an open type ([§8.4.3](types.md#843-open-and-closed-types)).  A declaration pattern naming a type `T` is *applicable to* every type `E` for which `E` is *pattern compatible* with `T`.

> *Note*: The support for open types can be most useful when checking types that may be either struct or class types, and boxing is to be avoided. *end note*
<!-- markdownlint-disable MD028 -->
<!-- markdownlint-enable MD028 -->
> *Example*: The declaration pattern is useful for performing run-time type tests of reference types, and replaces the idiom
>
> ```csharp
> var v = expr as Type;
> if (v != null) { /* code using v */ }
> ```
>
> with the slightly more concise
>
> ```csharp
> if (expr is Type v) { /* code using v */ }
> ```
>

> > *end example*

Author: ```suggestion
> *end example*

```

> It is an error if *type* is a nullable value type.

Reviewer: Missing blank line after `> *end example*`

> It is an error if *type* is a nullable value type.

Author: ```suggestion

It is an error if *type* is a nullable value type.
```

> > *Example*: The declaration pattern can be used to test values of nullable types: a value of type `Nullable<T>` (or a boxed `T`) matches a type pattern `T2 id` if the value is non-null and the type of `T2` is `T`, or some base type or interface of `T`. For example, in the code fragment

Reviewer: ```suggestion
> *Example*: The declaration pattern can be used to test values of nullable types: a value of type `Nullable<T>` (or a boxed `T`) matches a type pattern `T2 id` if the value is non-null and `T2` is `T`, or some base type or interface of `T`. For example, in the code fragment
```

>
> ```csharp
> int? x = 3;
> if (x is int v) { /* code using v */ }
> ```
>
> The condition of the `if` statement is `true` at runtime and the variable `v` holds the value `3` of type `int` inside the block. *end example*

### §constant-pattern-new-clause Constant pattern

A *constant_pattern* is used to test the value of a pattern input value (§patterns-new-clause) against the given constant value.

```ANTLR
constant_pattern
    : constant_expression
    ;
```

A constant pattern `P` is *applicable to* a type `T` if there is an implicit conversion from the constant expression of `P` to the type `T`.

For a constant pattern `P`, we say its *converted value* is

> - if the input expression's type is an integral type or an enum type, the pattern's constant value converted to that type; otherwise

Author: ```suggestion

- if the input expression's type is an integral type or an enum type, the pattern's constant value converted to that type; otherwise
```

- if the input expression's type is the nullable version of an integral type or an enum type, the pattern's constant value converted to its underlying type; otherwise
- the value of the pattern's constant value.

Given a *pattern input value* *e* and a a constant pattern `P` with converted value *v*,

> - if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise

Author: ```suggestion

- if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise
```

    ;
```

Given a *pattern input value* *e*, if *designation* is the *identifier* `_`, it denotes a discard (§discards-new-clause), and the value of *e* is not bound to anything. (Although a declared variable with that name may be in scope at that point, that named variable is not seen in this context.) If *designation* is any other identifier, at runtime the value of *e* is bound to a newly introduced local variable ([§9.2.8](variables.md#928-local-variables)) of that name whose type is the static type of *e*, and the pattern input value is assigned to that local variable.

It is an error if the name `var` would bind to a type where a *var_pattern* is used.

## Pattern Subsumption

In a switch statement, it is an error if a case's pattern is *subsumed* by the preceding set of unguarded cases (XREF).
Informally, this means that any input value would have been matched by one of the previous cases.
Here we define when a set of patterns *subsumes* a given pattern.

We say a pattern `P` *would match* a constant `K` if the specification for that pattern's runtime behavior is that `P` matches `K`.

A set of patterns `Q` *subsumes* a pattern `P` if any of the following conditions hold:

> - `P` is a constant pattern and any of the patterns in the set `Q` would match `P`'s *converted value*

Author: ```suggestion

- `P` is a constant pattern and any of the patterns in the set `Q` would match `P`'s *converted value*
```

- `P` is a var pattern and the set of patterns `Q` is *exhaustive* for the type of the pattern input value, and either the pattern input value is not of a nullable type or some pattern in `Q` would match `null`.
- `P` is a declaration pattern with type `T` and the set of patterns `Q` is *exhaustive* for the type `T` (XREF).

## Pattern Exhaustiveness

Informally, we say that a set of patterns is exhaustive for a type if some pattern in the set is applicable to every possible value of that type other than null.
Here we define when a set of patterns is *exhaustive* for a type.

> A set of patterns `Q` is *exhaustive* for a type `T` if any of the following conditions hold:

Author: ```suggestion
A set of patterns `Q` is *exhaustive* for a type `T` if any of the following conditions hold:

```

> 1. `T` is an integral or enum type, or a nullable version of one of those, and for every possible value of `T`'s underlying type, some pattern in `Q` would that value; or

Reviewer: ```suggestion
1. `T` is an integral or enum type, or a nullable version of one of those, and for every possible value of `T`'s underlying type, some pattern in `Q` would match that value; or
```

### §constant-pattern-new-clause Constant pattern

A *constant_pattern* is used to test the value of a pattern input value (§patterns-new-clause) against the given constant value.

```ANTLR
constant_pattern
    : constant_expression
    ;
```

A constant pattern `P` is *applicable to* a type `T` if there is an implicit conversion from the constant expression of `P` to the type `T`.

For a constant pattern `P`, we say its *converted value* is

- if the input expression's type is an integral type or an enum type, the pattern's constant value converted to that type; otherwise
- if the input expression's type is the nullable version of an integral type or an enum type, the pattern's constant value converted to its underlying type; otherwise
- the value of the pattern's constant value.

Given a *pattern input value* *e* and a constant pattern `P` with converted value *v*,

- if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise
- the pattern `P` *matches* the value *e* if `object.Equals(e, v)` returns `true`.

> > *Example*:

Reviewer 4: ```suggestion
> *Example*: The `switch` statement in the following method uses five constant patterns in its case labels.
```

Author: Should be done sooner rather than later.

I would suggest placing this new chapter between the existing 12 (Statements) and 13 (namespaces)

# §patterns-new-clause Patterns and pattern matching

## §patterns-new-clause-general General

A ***pattern*** is a syntactic form that can be used with the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)) and in a *switch_statement* ([§12.8.3](statements.md#1283-the-switch-statement)) to express the shape of data against which incoming data is to be compared. A pattern is tested against the *expression* of a switch statement, or against a *relational_expression* that is on the left-hand side of an `is` operator. We call this a ***pattern input value***.

## §patterns-new-clause-forms Pattern Forms

A pattern may have one of the following forms:

```ANTLR
pattern
    : declaration_pattern
    | constant_pattern
    | var_pattern
    ;
```

> A *declaration_pattern* and a *var_pattern* can result in the declaration of a local variable.

Reviewer 4: *Can* result, or *always* result (but possibly without the variable being definitely assigned)?

Reviewer 4: Ah - discards, right?

Author: Right.

> Each pattern form defines the set of types for input values that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.

Reviewer 4: I think we may want an example here to clarify this. I believe I get it, but it could be confusing.

For example, consider:

```csharp
if (x is string s)
```

It's fine for the compile-time type of `x` to be `object`, even though a value whose execution type is `object` *won't* match. But the expression type being `object` is okay because the value *might* be a `string` reference. Whereas if the compile-time type of `x` is (say) `Stream` then this would be a compile-time error.

I think that sort of clarification can be added later - I don't want us to wait for perfection before merging this, so long as the text is accurate when read appropriately (and I think it is here). We might want to keep a list of this sort of thing somewhere, so we can come back afterwards. (It's inevitable when adding a huge chunk this PR does.)

Author: Mention this is a compile-time concept.

The simplest way to do that is change "it is an error" to "it is a compile-time error".

Author: Agreed, that expanding upon this, perhaps with a sentence or two just as you wrote here, is worth doing (later)

Reviewer 3: "compile time" added in #934 

#938 logs the request for an example.

> Each pattern form defines the set of values for which the pattern *matches* the value.

Author: Say this is a runtime concept.  (Not urgent, or even necessary)

### §declaration-pattern-new-clause Declaration pattern

> A *declaration_pattern* is used to test that a value has a given type and, if the test succeeds, provide the value in a variable of that type.

Reviewer 2: From below its "a value has a type which is pattern compatible with a given type" – which is a bit long winded so some wordsmithing is required…

Author: No, this is a different, runtime concept.

(no action needed)

Author: Should take its description from the is-type operator.

Best addressed sooner rather than later.

Author: This would be better done sooner rather than later.  Two things:
1. The specification about the value being of the given type is actually a bit more complex and should delegate to the spec of the is-type operator.  This is a runtime type test.
2. The *pattern compatible* concept is to give warnings based on compile-time types that are impossible to match.  If that isn't clear, it should be made clear.

Reviewer 3: Addressed in #934

Added text that the runtime test must produced the same results as the is-type operator. Earlier edits should have already clarified the *pattern compatible* concept as a compile-time operation.

```ANTLR
declaration_pattern
    : type simple_designation
    ;
simple_designation
    : single_variable_designation
    ;
single_variable_designation
    : identifier
    ;
```

The runtime type of the value is tested against the *type* in the pattern. If it is of that runtime type (or some subtype), the pattern *matches* that value. This pattern form never matches a `null` value.

> Given a *pattern input value* *e*, if the *simple_designation* is the *identifier* `_`, it denotes a discard (§discards-new-clause) the value of *e* is not bound to anything. (Although a declared variable with the name `_` may be in scope at that point, that named variable is not seen in this context.) If *simple_designation* is any other identifier, a local variable ([§9.2.8](variables.md#928-local-variables)) of the given type named by the given identifier is introduced. That local variable is assigned the value of the *pattern input value* when the pattern *matches* the value.

Reviewer 4: I suspect the rendering of *pattern input value* *e* would be clearer without "pattern input value" in italics. Or we could try to reword, something like "Given a *pattern input value* of the form *e*" - although that implies there are other forms. Thoughts welcome.

Author: I'm open to seeing the wording reconsidered.  Also, the type matching is more than just subtype.  I believe it should delegate the specification to the `is-type` operator (e.g. include boxing and unboxing).  Not urgent.

Reviewer 3: Already addressed in earlier edits in #934

> Certain combinations of static type of the pattern input value and the given type are considered incompatible and result in a compile-time error. A value of static type `E` is said to be ***pattern compatible*** with the type `T` if there exists an identity conversion, an implicit reference conversion, a boxing conversion, an explicit reference conversion, or an unboxing conversion from `E` to `T`, or if either `E` or `T` is an open type ([§8.4.3](types.md#843-open-and-closed-types)).  A declaration pattern naming a type `T` is *applicable to* every type `E` for which `E` is *pattern compatible* with `T`.

Reviewer 4: Can we reduce "an implicit reference conversion, a boxing conversion, an explicit reference conversion" to "a reference conversion, a boxing conversion, ..."?

Reviewer 4: Nit: the "period double-space" before "A declaration pattern" makes me twitch. I'm happy to change those afterwards :)

Author: > Can we reduce "an implicit reference conversion, a boxing conversion, an explicit reference conversion" to "a reference conversion, a boxing conversion, ..."?

We don't define the phrase "reference conversion".  However, we could say "implicit or explicit reference conversion".  However, this is super low priority.

> Nit: the "period double-space" before "A declaration pattern" makes me twitch. I'm happy to change those afterwards :)

One of us will end up twitching.

Reviewer 3: I added the change to #934 (because it was even easier than logging the low priority issue).

I checked and removed the remaining instances of "period double-space" exist in any clause, except where the two spaces was used at the end of a line to force a line break.

> *Note*: The support for open types can be most useful when checking types that may be either struct or class types, and boxing is to be avoided. *end note*
<!-- markdownlint-disable MD028 -->
<!-- markdownlint-enable MD028 -->
> *Example*: The declaration pattern is useful for performing run-time type tests of reference types, and replaces the idiom
>
> ```csharp
> var v = expr as Type;
> if (v != null) { /* code using v */ }
> ```
>
> with the slightly more concise
>
> ```csharp
> if (expr is Type v) { /* code using v */ }
> ```
>
> *end example*

> It is an error if *type* is a nullable value type.

Author: ... unless it is the type of an is-type expression?

> *Example*: The declaration pattern can be used to test values of nullable types: a value of type `Nullable<T>` (or a boxed `T`) matches a type pattern `T2 id` if the value is non-null and `T2` is `T`, or some base type or interface of `T`. For example, in the code fragment
>
> ```csharp
> int? x = 3;
> if (x is int v) { /* code using v */ }
> ```
>
> The condition of the `if` statement is `true` at runtime and the variable `v` holds the value `3` of type `int` inside the block. *end example*

### §constant-pattern-new-clause Constant pattern

A *constant_pattern* is used to test the value of a pattern input value (§patterns-new-clause) against the given constant value.

```ANTLR
constant_pattern
    : constant_expression
    ;
```

A constant pattern `P` is *applicable to* a type `T` if there is an implicit conversion from the constant expression of `P` to the type `T`.

For a constant pattern `P`, we say its *converted value* is

> - if the input expression's type is an integral type or an enum type, the pattern's constant value converted to that type; otherwise

Author: not called input expression...

Author: Change "input expression" to "pattern input value".

Reviewer 3: Edited. Check for any additional occurrence of "input expression" where "pattern input value" should be used. None found.

- if the input expression's type is the nullable version of an integral type or an enum type, the pattern's constant value converted to its underlying type; otherwise
- the value of the pattern's constant value.

> Given a *pattern input value* *e* and a a constant pattern `P` with converted value *v*,

Reviewer 4: ```suggestion
Given a *pattern input value* *e* and a constant pattern `P` with converted value *v*,
```

- if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise

> - the pattern `P` *matches* the value *e* if `object.Equals(e, v)` returns `true`.

Reviewer 4: It looks like the compiler is allowed to generate code with the same effect as that, but without actually calling `object.Equals` (which will box for float/decimal/double of course). For example, I've just tried `d is double.NaN` and Roslyn compiles that to `double.IsNaN(d)`.
Do we need to mention that?

Author: Generally, no.

> > *Example*:

Reviewer 4: I think this example could do with a bit more explanation - particularly as unless you're attuned to it already, you might be wondering where the pattern is.

(Also, please put the opening brace on a new line.)

Author: Please propose an edit.

Author: This minor typographical change can be deferred.

Reviewer 3: Added a sentence, and moved the opening brace.

>
> ```csharp
> public static decimal GetGroupTicketPrice(int visitorCount)
> {
>     switch (visitorCount) {
>         case 1: return 12.0m;
>         case 2: return 20.0m;
>         case 3: return 27.0m;
>         case 4: return 32.0m;
>         case 0: return 0.0m;
>         default: throw new ArgumentException(…);
>     }
> }
> ```
>
> *end example*

### §var-pattern-new-clause Var pattern

> A *var_pattern* matches every value.  That is, a pattern-matching operation with a *var_pattern* always succeeds.

Author: Matches in italics

Author: Sooner, please.

Reviewer 3: Added in #934

It is an error if the name `var` would bind to a type where a *var_pattern* is used.

## Pattern Subsumption

In a switch statement, it is an error if a case's pattern is *subsumed* by the preceding set of unguarded cases (XREF).
Informally, this means that any input value would have been matched by one of the previous cases.
Here we define when a set of patterns *subsumes* a given pattern.

We say a pattern `P` *would match* a constant `K` if the specification for that pattern's runtime behavior is that `P` matches `K`.

A set of patterns `Q` *subsumes* a pattern `P` if any of the following conditions hold:

- `P` is a constant pattern and any of the patterns in the set `Q` would match `P`'s *converted value*
- `P` is a var pattern and the set of patterns `Q` is *exhaustive* for the type of the pattern input value, and either the pattern input value is not of a nullable type or some pattern in `Q` would match `null`.
- `P` is a declaration pattern with type `T` and the set of patterns `Q` is *exhaustive* for the type `T` (XREF).

## Pattern Exhaustiveness

> Informally, we say that a set of patterns is exhaustive for a type if some pattern in the set is applicable to every possible value of that type other than null.

Reviewer 4: There are two ways of reading "if some pattern in the set is applicable to every possible value of that type" - my initial reading is "one of the patterns in the set has the property that it matches every possible value of that type". Whereas I assume you really mean "if every possible value of that type is matched by at least one pattern in the set".

Let's add this to the "non-blocking, could do with tweaking" list - assuming I've read it correctly.

Author: Agreed.

Reviewer 3: Logged in #939

Here we define when a set of patterns is *exhaustive* for a type.

A set of patterns `Q` is *exhaustive* for a type `T` if any of the following conditions hold:

> 1. `T` is an integral or enum type, or a nullable version of one of those, and for every possible value of `T`'s underlying type, some pattern in `Q` would match that value; or

Author: "underlying" is meant to apply to nullable types, not enum types.

Author: change "underlying" to "nullable underlying".

Reviewer 3: added in #934
